File: backend-py/src/voxpilot/services/tools/grep_search.py
"""grep_search tool — search file contents by regex pattern."""

import re
from pathlib import Path
from typing import Any, ClassVar

from voxpilot.services.tools.base import Tool

# Directories to skip when walking
_SKIP_DIRS = {
    ".git",
    "__pycache__",
    "node_modules",
    ".venv",
    "venv",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
}

MAX_MATCHES = 200
MAX_LINE_LENGTH = 500


class GrepSearchTool(Tool):
    """Search file contents by regex pattern."""

    name = "grep_search"
    description = (
        "Search for a regex pattern in file contents within the working directory. "
        "Returns matching lines with file paths and line numbers. "
        "Optionally restrict to a subdirectory and/or glob file pattern."
    )
    parameters: ClassVar[dict[str, Any]] = {
        "type": "object",
        "properties": {
            "pattern": {
                "type": "string",
                "description": "Regular expression pattern to search for.",
            },
            "path": {
                "type": "string",
                "description": (
                    "Subdirectory to search within (relative to working directory). "
                    "Defaults to '.' (entire working directory)."
                ),
            },
            "include": {
                "type": "string",
                "description": (
                    "Glob pattern to filter files (e.g., '*.py', '*.ts'). "
                    "If omitted, searches all text files."
                ),
            },
        },
        "required": ["pattern"],
        "additionalProperties": False,
    }

    async def execute(self, arguments: dict[str, Any], work_dir: Path) -> str:
        """Search for a regex pattern in files."""
        pattern_str: str = arguments.get("pattern", "")
        if not pattern_str:
            return "Error: 'pattern' argument is required."

        try:
            regex = re.compile(pattern_str, re.IGNORECASE)
        except re.error as exc:
            return f"Error: invalid regex pattern '{pattern_str}': {exc}"

        raw_path: str = arguments.get("path", ".")
        resolved = self._resolve_path(raw_path, work_dir)
        if resolved is None:
            return f"Error: path '{raw_path}' is outside the working directory."
        if not resolved.exists():
            return f"Error: path '{raw_path}' does not exist."

        include: str | None = arguments.get("include")
        matches: list[str] = []
        files_searched = 0

        for file_path in self._walk_files(resolved, include):
            files_searched += 1
            try:
                text = file_path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue

            rel = file_path.relative_to(work_dir.resolve())
            for line_no, line in enumerate(text.splitlines(), start=1):
                if regex.search(line):
                    display = line[:MAX_LINE_LENGTH]
                    if len(line) > MAX_LINE_LENGTH:
                        display += "..."
                    matches.append(f"{rel}:{line_no}: {display}")
                    if len(matches) >= MAX_MATCHES:
                        matches.append(
                            f"... (truncated at {MAX_MATCHES} matches)"
                        )
                        return self._format_result(
                            pattern_str, matches, files_searched
                        )

        return self._format_result(pattern_str, matches, files_searched)

    def _walk_files(self, root: Path, include: str | None) -> list[Path]:
        """Collect files to search, respecting skip dirs and optional glob."""
        files: list[Path] = []
        if root.is_file():
            files.append(root)
            return files

        for entry in sorted(root.rglob("*")):
            # Skip noise directories
            if any(part in _SKIP_DIRS for part in entry.parts):
                continue
            if not entry.is_file():
                continue
            # Apply include glob filter
            if include and not entry.match(include):
                continue
            # Skip likely binary files
            if self._is_likely_binary(entry):
                continue
            files.append(entry)
        return files

    @staticmethod
    def _is_likely_binary(path: Path) -> bool:
        """Heuristic: skip files with common binary extensions."""
        binary_exts = {
            ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico", ".svg",
            ".woff", ".woff2", ".ttf", ".eot",
            ".zip", ".tar", ".gz", ".bz2", ".xz", ".7z",
            ".exe", ".dll", ".so", ".dylib", ".o", ".a",
            ".pyc", ".pyo", ".class", ".jar",
            ".db", ".sqlite", ".sqlite3",
            ".pdf", ".doc", ".docx", ".xls", ".xlsx",
            ".mp3", ".mp4", ".avi", ".mov", ".wav",
        }
        return path.suffix.lower() in binary_exts

    @staticmethod
    def _format_result(
        pattern: str, matches: list[str], files_searched: int
    ) -> str:
        if not matches:
            return f"No matches found for pattern '{pattern}' ({files_searched} files searched)."
        header = (
            f"Found {len(matches)} match(es) for '{pattern}'"
            f" ({files_searched} files searched):\n"
        )
        return header + "\n".join(matches)

File: backend/src/tools/grep-search.ts
import { z } from "zod/v4";
import { readdir, readFile, stat } from "node:fs/promises";
import type { Dirent } from "node:fs";
import { join, relative, resolve, extname } from "node:path";
import { type Tool, type ToolResult, resolvePath, simpleResult } from "./base";

const SKIP_DIRS = new Set([
  ".git",
  "__pycache__",
  "node_modules",
  ".venv",
  "venv",
  ".mypy_cache",
  ".pytest_cache",
  ".ruff_cache",
]);

const MAX_MATCHES = 200;
const MAX_LINE_LENGTH = 500;

const BINARY_EXTENSIONS = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".bmp",
  ".ico",
  ".svg",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".zip",
  ".tar",
  ".gz",
  ".bz2",
  ".xz",
  ".7z",
  ".exe",
  ".dll",
  ".so",
  ".dylib",
  ".o",
  ".a",
  ".pyc",
  ".pyo",
  ".class",
  ".jar",
  ".db",
  ".sqlite",
  ".sqlite3",
  ".pdf",
  ".doc",
  ".docx",
  ".xls",
  ".xlsx",
  ".mp3",
  ".mp4",
  ".avi",
  ".mov",
  ".wav",
]);

const parameters = z
  .object({
    pattern: z
      .string()
      .describe("Regular expression pattern to search for."),
    path: z
      .string()
      .optional()
      .describe(
        "Subdirectory to search within (relative to working directory). " +
          "Defaults to '.' (entire working directory).",
      ),
    include: z
      .string()
      .optional()
      .describe(
        "Glob pattern to filter files (e.g., '*.py', '*.ts'). " +
          "If omitted, searches all text files.",
      ),
  })
  .strict();

type Params = z.infer<typeof parameters>;

export class GrepSearchTool implements Tool<typeof parameters> {
  readonly name = "grep_search";
  readonly description =
    "Search for a regex pattern in file contents within the working directory. " +
    "Returns matching lines with file paths and line numbers. " +
    "Optionally restrict to a subdirectory and/or glob file pattern.";
  readonly parameters = parameters;
  readonly requiresConfirmation = false;

  async execute(args: Params, workDir: string): Promise<ToolResult> {
    const patternStr = args.pattern;
    if (!patternStr) {
      return simpleResult("Error: 'pattern' argument is required.");
    }

    let regex: RegExp;
    try {
      regex = new RegExp(patternStr, "i");
    } catch (exc) {
      return simpleResult(
        `Error: invalid regex pattern '${patternStr}': ${exc}`,
      );
    }

    const rawPath =
      args.path && args.path !== "" ? args.path : ".";

    const resolved = await resolvePath(rawPath, workDir);
    if (resolved === null) {
      return simpleResult(
        `Error: path '${rawPath}' is outside the working directory.`,
      );
    }

    try {
      await stat(resolved);
    } catch {
      return simpleResult(`Error: path '${rawPath}' does not exist.`);
    }

    const include = args.include;
    const absWorkDir = resolve(workDir);
    const files = await this.walkFiles(resolved, include);
    const matches: string[] = [];
    let filesSearched = 0;

    for (const filePath of files) {
      filesSearched++;
      let text: string;
      try {
        text = await readFile(filePath, "utf-8");
      } catch {
        continue;
      }

      const rel = relative(absWorkDir, filePath);
      const lines = text.split("\n");
      for (let lineNo = 0; lineNo < lines.length; lineNo++) {
        const line = lines[lineNo];
        if (line !== undefined && regex.test(line)) {
          let display = line.slice(0, MAX_LINE_LENGTH);
          if (line.length > MAX_LINE_LENGTH) {
            display += "...";
          }
          matches.push(`${rel}:${lineNo + 1}: ${display}`);
          if (matches.length >= MAX_MATCHES) {
            matches.push(`... (truncated at ${MAX_MATCHES} matches)`);
            return simpleResult(
              this.formatResult(patternStr, matches, filesSearched),
            );
          }
        }
      }
    }

    return simpleResult(this.formatResult(patternStr, matches, filesSearched));
  }

  private async walkFiles(
    root: string,
    include: string | undefined,
  ): Promise<string[]> {
    const files: string[] = [];

    let rootStat: Awaited<ReturnType<typeof stat>>;
    try {
      rootStat = await stat(root);
    } catch {
      return files;
    }

    if (rootStat.isFile()) {
      files.push(root);
      return files;
    }

    const includeGlob = include ? new Bun.Glob(include) : undefined;
    await this.walkDir(root, files, includeGlob);
    files.sort();
    return files;
  }

  private async walkDir(
    dir: string,
    files: string[],
    includeGlob: InstanceType<typeof Bun.Glob> | undefined,
  ): Promise<void> {
    let entries: Dirent[];
    try {
      entries = await readdir(dir, { withFileTypes: true, encoding: "utf-8" });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (SKIP_DIRS.has(entry.name)) continue;
        await this.walkDir(join(dir, entry.name), files, includeGlob);
      } else if (entry.isFile()) {
        if (this.isLikelyBinary(entry.name)) continue;
        if (includeGlob && !includeGlob.match(entry.name)) continue;
        files.push(join(dir, entry.name));
      }
    }
  }

  private isLikelyBinary(name: string): boolean {
    return BINARY_EXTENSIONS.has(extname(name).toLowerCase());
  }

  private formatResult(
    pattern: string,
    matches: string[],
    filesSearched: number,
  ): string {
    if (matches.length === 0) {
      return `No matches found for pattern '${pattern}' (${filesSearched} files searched).`;
    }
    const header = `Found ${matches.length} match(es) for '${pattern}' (${filesSearched} files searched):\n`;
    return header + matches.join("\n");
  }
}
